Use enc: dir (#7)
ietf-wg-jose/draft-ietf-jose-hpke-encrypt: draft-ietf-jose-hpke-encrypt.md

In HPKE JWE Integrated Encryption, JWE Encrypted Key is the encapsulated key.

In HPKE JWE Key Encryption, each recipient JWE Encrypted Key is the encrypted content encryption key, and the encapsulated key (ek) is found in the recipient header.

# Integrated Encryption

In HPKE JWE Integrated Encryption:

- The protected header MUST contain an "alg" that starts with "HPKE".

> - The protected header MUST contain an "enc" and it must be set to the value "dir".

Reviewer: You may want to replace "must" with "MUST".

- The protected header parameters "psk_id" and "auth_kid" MAY be present.
- The protected header parameters "ek" MUST NOT be present.
- The "encrypted_key" MUST be the base64url encoded encapsulated key as defined in Section 5.1.1 of {{RFC9180}}.
- The "iv", "tag" and "aad" members MUST NOT be present.
- The "ciphertext" MUST be the base64url encoded ciphertext as defined in Section 5.2 of {{RFC9180}}.

> - The HPKE Setup info parameter MAY be used and its values are not constrained by this specification, by default it is empty.

Reviewer: we will have to remove the following line from the draft: 
The "Setup info" MUST NOT be used with either HPKE JWE Integrated Encryption and HPKE JWE Key Encryption.

Reviewer: I think for interoperability we should say the following for both modes :
When apu / apv are present, the HPKE Setup info parameter will carry the JOSE context-specific data defined in Section 4.6.2 of [RFC7518].

## Compact Example

A Compact JWE or JSON Web Token:

~~~
eyJhbGciOiJIUEtFLVAyNTYtU0hBMjU2LUExMjhHQ00iLCJlbmMiOiJkaXIiLCJraWQiOiJ1cm46aWV0ZjpwYXJhbXM6b2F1dGg6andrLXRodW1icHJpbnQ6c2hhLTI1Njp2b2RIQ3FjVVdFbV83NUpWcXlhTjhaS1FVMjF3VEFSYzhkRzhuVU1jZlBVIn0.BCsvYxTHM4CO_OwQxL3lkJDdlw3UDjx2xN9MIXnbVzfTgFJmo_Es2xdH-fYs9EXfH_V53JgMWfUm7rBD_oE5efU..7_va6cnwClMsw7h7lqpm2tCrH9NkciM-g9UabdPWcOeIRmAf01NLYG7Wn8fFoohHlcGgd0nh7Jmo9nvHFi7sH6kOX7pplBnvLUoPrqeyW4TdXo_X8YThNKf9BFyWGyF6fjelbic5jSYClFaenMkTnjpHxFW1sWuiuZVmO1EOzrlNttWy.
~~~

After verification:

~~~
{
  "protectedHeader": {

>     "alg": "HPKE-P256-SHA256-A128GCM",

Reviewer 2: We should deal with this in a different issue, but this is the longest "alg" value ever - by far.  They're meant to be short registered identifiers - not strings you can parse.

- The protected header MUST contain an "enc" that is registered in both the IANA HPKE AEAD Identifiers Registry, and the IANA JSON Web Signature and Encryption Algorithms Registry.
- The recipient unprotected header parameters "psk_id" and "auth_kid" MAY be present.
- The recipient unprotected header parameter "ek" MUST be present.

> - The recipient unprotected header MUST contain an "alg" that starts with "HPKE".

Reviewer 2: ```suggestion
- The recipient unprotected header MUST contain a registered HPKE "alg" value.
```

> - The recipient unprotected header MUST contain an "enc" and it must be set to the value "dir".

Reviewer: In the case of Key Encryption, what is the need to convey "enc" in the recipient unprotected header? I don’t see a need for it in this mode, as the HPKE JWE Key Encryption Mode can function similarly to Key Agreement with ECDH-ES.
